packages/fpdart/lib/src/effect.dart

```diff
@@ -0,0 +1,800 @@
+              onError: (error, stackTrace) {
+                if (error is _EffectThrow<L>) {
+                  return Left<Cause<L>, R>(error.cause);
+                }
+
+                return Left<Cause<L>, R>(Die(error, stackTrace));
+              },
+            );
+          } on _EffectThrow<L> catch (genError) {
+            return Left(genError.cause);
+          }
+        },
+      );
+
+  /// {@category constructors}
+  factory Effect.tryCatch({
+    required FutureOr<R> Function() execute,
+    required L Function(Object error, StackTrace stackTrace) onError,
+    FutureOr<dynamic> Function()? onCancel,
+  }) =>
+      Effect.from(
+        (env) {
+          try {
+            return execute().then(
+              Right.new,
+              onError: (error, stackTrace) => Left(
+                Failure(onError(error, stackTrace), stackTrace),
+              ),
+            );
+          } catch (err, stack) {
+            return Left(Failure(onError(err, stack), stack));
+          }
+        },
+      );
+
+  /// {@category constructors}
+  factory Effect.fromNullable(R? value, {required L Function() onNull}) =>
+      Effect.from(
+        (_) => value == null ? Left(Failure(onNull())) : Right(value),
+      );
+
+  /// {@category constructors}
+  factory Effect.functionFail(FutureOr<Cause<L>> Function() f) => Effect.from(
+        (_) => f().then(Left.new),
+      );
+
+  /// {@category constructors}
+  factory Effect.functionSucceed(FutureOr<R> Function() f) => Effect.from(
+        (_) => f().then(Right.new),
+      );
+
+  /// {@category constructors}
+  factory Effect.lazy(Effect<E, L, R> Function() effect) =>
+      Effect.from((context) => effect().__unsafeRun(context));
+
+  /// {@category constructors}
+  factory Effect.fail(L value) => Effect.from((_) => Left(Failure(value)));
+
+  /// {@category constructors}
+  factory Effect.failCause(Cause<L> cause) => Effect.from((_) => Left(cause));
+
+  /// {@category constructors}
+  factory Effect.succeed(R value) => Effect.from((_) => Right(value));
+
+  /// {@category constructors}
+  factory Effect.raceAll(Iterable<Effect<E, L, R>> iterable) =>
+      Effect.from((context) {
+        final signal = Deferred<Never, Never>.unsafeMake();
+        final deferred = Deferred<L, R>.unsafeMake();
+
+        for (final effect in iterable) {
+          effect
+              .__unsafeRun(context.withSignal(signal))
+              .then(deferred.unsafeCompleteExit);
+
+          if (deferred.unsafeCompleted) {
+            break;
+          }
+        }
+
+        return deferred.wait<E>().__unsafeRun(context).then(
+              (exit) => signal
+                  .failCause<E, L>(const Interrupted())
+                  .__unsafeRun(context.withoutSignal)
+                  .then((_) => exit),
+            );
+      });
+
+  /// {@category constructors}
+  static Effect<E, L, void> sleep<E, L>(Duration duration) => Effect.from(
+        (_) => Future.delayed(
```

Comment:
Maybe dart has a lower level timer api that can be interrupted, to prevent resource leaks.